src/lib/uri/uri.h

```diff
@@ -49,6 +49,23 @@ struct uri_set {
 #define URI_MAXHOST NI_MAXHOST
 #define URI_MAXSERVICE _POSIX_PATH_MAX /* _POSIX_PATH_MAX always > NI_MAXSERV */
 
+/**
+ * Add new @a value to URI query @a param.
+ */
+void
+uri_query_param_add_value(struct uri_query_param *param, const char *value);
+
+/**
+ * Add new query parameter to @uri structure with given @a name.
+ */
+struct uri_query_param *
+uri_add_query_param(struct uri *uri, const char *name);
+
+/**
+ * Find URI query parameter by its @a name in @a uri structure.
+ */
+struct uri_query_param *
+uri_find_query_param(const struct uri *uri, const char *name);
```

Comment:
I don't really like that we have to make the `uri_query_param` entity public. Maybe, add two `uri` methods instead?

```C
// Appends a value to a uri query parameter, creating one if it doesn't exist.
void uri_add_query_param(struct uri *uri, const char *name, const char *value);
// Removes a uri query parameter and all its values.
void uri_remove_query_param(struct uri *uri, const char *name);
```

This would be slower, because we would have to look up a query parameter per each insertion, but we don't care about it when we look up query parameter values:

```C
int count = uri_query_param_count(my_uri, "my_param"); // look up
for (int i = 0; i < count; i++) {
    const char *value = uri_query_param(my_uri, "my_param", i); // another look up
    // ...
}
```

Let's be consistent: either use `uri_query_param` struct both for lookups and modifications or use the slow but simple API everywhere. You should discuss this simplicity-vs-performance issue with @alyapunov and @Gerold103. Personally, I don't think we need to be fast here, because the number of uri parameters is expected to be low; besides we can optimize out the lookups in future by caching the last accessed parameter.

Reply:
fixed